File: we_jump_dotnet/WeJump.Api/Utils/Charge.cs
namespace WeJump.Api.Utils;

/// <summary>
/// 蓄力时长 -> 步数的映射（服务端权威，与客户端保持一致；开局面板会把参数下发给客户端）。
/// 第 k 格蓄满耗时 = FirstCellMs * Ratio^(k-1) ms；蓄满 k 格累计耗时 C(k) = FirstCellMs * (1 - Ratio^k)/(1 - Ratio) ms。
/// </summary>
public static class Charge
{
    /// <summary>单跳最大格数上限（全局，两端一致），由服务端下发给客户端。</summary>
    public const int MaxStep = 10;

    /// <summary>第 1 格蓄满耗时(ms)。</summary>
    public const double FirstCellMs = 300.0;

    /// <summary>逐格衰减系数。</summary>
    public const double Ratio = 0.9;

    /// <summary>蓄满 k 格所需累计时长(ms)，k>=1。</summary>
    public static double CumTimeMs(int k)
    {
        double acc = 0;
        for (int i = 1; i <= k; i++)
            acc += FirstCellMs * Math.Pow(Ratio, i - 1);
        return acc;
    }

    /// <summary>给定蓄力时长(ms)反推步数：能充满几格就算几步，至少 1 步。</summary>
    public static int StepsForMs(double ms)
    {
        if (ms < 0) return 1;
        int s = 1;
        for (int k = 1; k <= MaxStep; k++)
        {
            if (ms >= CumTimeMs(k) - 1) s = k;
            else break;
        }
        return s;
    }
}

File: we_jump_wx/js/logic/charge.js
/**
 * 蓄力规则与时长 -> 步数映射。
 *
 * 参数全部由服务端下发（`game_start` / `game_state` 快照里的 maxStep / firstCellMs / ratio），
 * 客户端不再内置任何规则常量，避免两端不一致导致“显示蓄了 N 格、服务端只结算 M 格”。
 * 本地映射仅用于进度环渲染与手感提示，最终结算始终以服务端为准。
 *
 * 公式：第 k 格蓄满耗时 = firstCellMs * ratio^(k-1)
 *      蓄满 k 格累计耗时 C(k) = firstCellMs * (1 - ratio^k) / (1 - ratio)
 */
var rules = {
  maxStep: 1,
  firstCellMs: 0,
  ratio: 0
};

/** 用服务端下发的参数刷新规则（缺失或非法的字段保留原值）。 */
function apply(src) {
  if (!src) return;
  if (src.maxStep > 0) rules.maxStep = src.maxStep;
  if (src.firstCellMs > 0) rules.firstCellMs = src.firstCellMs;
  if (src.ratio > 0) rules.ratio = src.ratio;
}

/** 单跳最大格数（即当前可蓄力上限）。 */
function maxStep() {
  return rules.maxStep > 0 ? rules.maxStep : 1;
}

/** 蓄满 k 格的累计时长(ms)。k>=1；未拿到参数时返回 Infinity。 */
function cumTime(k) {
  if (!(rules.firstCellMs > 0) || !(rules.ratio > 0)) return Infinity;
  var acc = 0;
  for (var i = 1; i <= k; i++) {
    acc += rules.firstCellMs * Math.pow(rules.ratio, i - 1);
  }
  return acc;
}

/** 给定蓄力时长(ms)与当前可跳上限 n，返回应跳步数（1..min(n, maxStep)）。 */
function stepsForMs(ms, n) {
  var limit = maxStep();
  var max = (n > 0 ? Math.min(n, limit) : limit);
  if (ms < 0 || !isFinite(cumTime(1))) return 1;
  var s = 1;
  for (var k = 1; k <= max; k++) {
    if (ms >= cumTime(k) - 1) s = k; else break;
  }
  return s;
}

module.exports = {
  apply: apply,
  maxStep: maxStep,
  cumTime: cumTime,
  stepsForMs: stepsForMs
};
